dspy/evaluate/evaluate.py

```diff
@@ -94,45 +93,61 @@ def _execute_multi_thread(self, wrapped_program, devset, num_threads, display_pr
         reordered_devset = []
         job_cancelled = "cancelled"
 
-        # context manger to handle sigint
-        @contextlib.contextmanager
-        def interrupt_handler_manager():
-            """Sets the cancel_jobs event when a SIGINT is received."""
-            default_handler = signal.getsignal(signal.SIGINT)
-
-            def interrupt_handler(sig, frame):
-                self.cancel_jobs.set()
-                dspy.logger.warning("Received SIGINT. Cancelling evaluation.")
-                default_handler(sig, frame)
-
-            signal.signal(signal.SIGINT, interrupt_handler)
-            yield
-            # reset to the default handler
-            signal.signal(signal.SIGINT, default_handler)
-
         def cancellable_wrapped_program(idx, arg):
             # If the cancel_jobs event is set, return the cancelled_job literal
             if self.cancel_jobs.is_set():
                 return None, None, job_cancelled, None
             return wrapped_program(idx, arg)
 
-        with ThreadPoolExecutor(max_workers=num_threads) as executor, interrupt_handler_manager():
-            futures = {executor.submit(cancellable_wrapped_program, idx, arg) for idx, arg in devset}
-            pbar = tqdm.tqdm(total=len(devset), dynamic_ncols=True, disable=not display_progress)
-
-            for future in as_completed(futures):
-                example_idx, example, prediction, score = future.result()
-
-                # use the cancelled_job literal to check if the job was cancelled - use "is" not "=="
-                # in case the prediction is "cancelled" for some reason.
-                if prediction is job_cancelled:
-                    continue
-
-                reordered_devset.append((example_idx, example, prediction, score))
-                ncorrect += score
-                ntotal += 1
-                self._update_progress(pbar, ncorrect, ntotal)
-            pbar.close()
+        try:
+            with ThreadPoolExecutor(max_workers=num_threads) as executor:
+                futures = {executor.submit(cancellable_wrapped_program, idx, arg) for idx, arg in devset}
+                pbar = tqdm.tqdm(total=len(devset), dynamic_ncols=True, disable=not display_progress, file=sys.stdout)
+
+                for future in as_completed(futures):
+                    if self.cancel_jobs.is_set():
+                        break  # Exit the loop if jobs are cancelled
+                    try:
+                        example_idx, example, prediction, score = future.result()
+                    except KeyboardInterrupt:
+                        self.cancel_jobs.set()
+                        dspy.logger.warning("Received KeyboardInterrupt. Cancelling evaluation.")
+                        # Cancel remaining futures
+                        for f in futures:
+                            f.cancel()
+                        break
```

Comment:
do you need to re-raise the KeyboardInterrupt here? If there's some other logic following on from `Evaluate` does the process get stopped or do you have to fire the interrupt again?